src/pam_rpasswd.8
.\" -*- nroff -*-
.\" Copyright (C) 2004, 2010 Thorsten Kukuk
.\" Author: Thorsten Kukuk <kukuk@suse.de>
.\"
.\" This program is free software; you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License version 2 as
.\" published by the Free Software Foundation.
.\"
.\" This program is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public License
.\" along with this program; if not, write to the Free Software Foundation,
.\" Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
.\"
.TH pam_rpasswd 8 "February 2004" "pwdutils"
.SH NAME
pam_rpasswd - PAM module to change remote password
.SH DESCRIPTION
The
.B pam_rpasswd
PAM module is for changing the password of user accounts
on a remote server over a secure SSL connection. It only
provides functionality for one PAM management group:
\fIpassword changing\fR.
.PP
The options needs to be added in the PAM configuration files.
.SH OPTIONS
.TP
.BI "config_file=" "<config-file>"
Specify another configuration file than \fI/etc/rpasswd.conf\fR.
.TP
.B "debug"
Print additinal debug messages to
.BR syslogd (8).
.TP
.BI "use_slp=" "0|1"
If compiled with SLP support, \fBpam_rpasswd\fR will use/not use
SLP to find a server. The default is to not use SLP.
.TP
.BI "reqcert=" "never|allow|try|hard"
.RS
Specifies what checks to perform on server certificates in
a SSL session. The default for this option is
.IR hard .
.TP
.B never
\fIpam_rpasswd\fR will not request or check any server certificate.
.TP
.B allow
The server certificate is requested. If no certificate is provided,
the session proceeds normally. If a bad certificate is provided, a
warning will be shown but the session proceeds normally.
.TP
.B try
The server certificate is requested. If no certificate is provided,
the session proceeds normally. If a bad certificate is provided,
the session is immediately terminated.
.TP
.B hard
The server certificate is requested. If no or a bad certificate is
provided, the session is immediately terminated. This is the default.
.RE
.TP
.BI "host=" "<hostname>"
Specify the name of the server, where
.BR rpasswdd (8)
is running.
.TP
.BI "port=" "<portnumber>"
Specify the portnumber on which the server process
.BR rpasswdd (8)
is listening.
.TP
.B "verbose"
Be verbose, print SSL connection data.
.TP
.B "quiet"
Be quiet, don't print connection trys or SSL connection data. This option
takes precedence over
.BR verbose .
.SH FILES
/etc/rpasswd.conf \- configuration file
.SH BUGS
This manual page is far from complete, most options are missing.
Please read the README and look at the source package.
.SH "SEE ALSO"
.BR login (1),
.BR passwd (1),
.BR pam.conf (8),
.BR pam.d (8),
.BR pam_pwcheck (8),
.BR pam (8),
.BR rpasswd (1),
.BR rpasswdd (8),
.BR rpc.yppasswdd (8),
.BR yppasswd (1)


NEWS
pwdutils NEWS -- history of user-visible changes.

Copyright (C) 2002-2012 Thorsten Kukuk

Please send bug reports, questions and suggestions to <kukuk@thkukuk.de>.

Version 3.2.19
* Fix passwd -S

Version 3.2.18
* Fix changing of permissions if uid/gid of user will be modified

Version 3.2.17
* Fix useradd.local

Version 3.2.16
* Make sha512 new default password hash

Version 3.2.15
* newgrp: honour dynamically assigned groups

Version 3.2.14
* pwck: check +/- entries, too
* grpck: check */- entries, too
* login.defs: remove LOGIN_RETRIES
 
Version 3.2.13
* Only copy regular files if usermod moves home directory

Version 3.2.12
* Make sure modified group files are written to disk
* Update translations for sv, fr

Version 3.2.11
* Update translations
* Add translation for nl

Version 3.2.10
* Fix encoding of polish translations
* Fix long options of grpck/pwck
 
Version 3.2.9
* Move UMASK from /etc/login.defs to /etc/default/useradd

Version 3.2.8
* Fix usermod manual page
* Add quiet option to pam_rpasswd.so

Version 3.2.7
* Fix possible endless loop in userdel

Version 3.2.6
* Fix rpasswd protocol error
* Update translations

Version 3.2.5
* Add translations for da
* pwconv: Remove temporary files in case of an error

Version 3.2.4
* Make sure changes are really written to disk

Version 3.2.3
* New translations
* Adjust documentation

Version 3.2.2
* Don't lock system accounts by default

Version 3.2.1
* Adjust for new libxcrypt

Version 3.2.0
* Better support for audit plugin
* --preferred-uid/gid option for useradd and groupadd

Version 3.1.4
* Update polish translation.
* Remove FTMP references.
* Better pwconv support for  NIS entries

Version 3.1.3
* Allow compiling without native language support.
* Update swedish translation.

Version 3.1.2
* Check if crontab is installed before calling it.
* Fix seg.fault in pwconv in error case.

Version 3.1.1
* Adjust PAM config files

Version 3.1.0
* gpasswd: Add -l and -u option to lock/unlock group passwords
* passwd: Add support for BioAPI
* Add LDAP support for RFC2307bis

Version 3.0.7
* passwd.1: correct manual page
* passwd: Allow changing passwords with PAM even if service is unknown

Version 3.0.6
* chfn: Fix syntax check for home phone number

Version 3.0.5
* Fix compilation error of PAM module with new Linux-PAM release
* Fix CHARACTER_CLASS in /etc/login.defs

Version 3.0.4
* Fix pam_rpasswd.so

Version 3.0.3
* usermod: Adjust sp_lstchg if -p option is used
* useradd: Preserve existing home directory
* chage: Fix off-by-one errors in output

Version 3.0.2
* Fix loading of logging plugins
* Unify options (use always --service)
* Add -a option to usermod for shadow compatibility

Version 3.0.1
* usermod: Don't remove old home directory if an error occured

Version 3.0
* Add --stdin option for passwd
* Fix useradd -D option
* Fix PAM/syslog interaction
* Fix lot of ignored failing syscalls
* Compile daemons with -fpie/-pie
* Add support for logging plugins
* Add plugin for LAuS support
* Add plugin for audit support
* Add plugin to log with syslog
* Use libnscd
* Add support for GnuTLS
* groupmod: Add options to remove/add single user from a group
* usermod: Add options to remove/add single user from a list of groups
* rpasswd/rpasswdd: Better SLP support

Version 2.6.4
* rpasswd/pam_rpasswd.so: Fix SLP URL parsing

Version 2.6.3
* newgrp: Add -c option to allow sg emulation

Version 2.6.2
* Fix compile problem if SELinux is disabled
* rpasswdd: Re-register SLP service after timeout

Version 2.6.1
* newgrp: Fix NGROUPS_MAX handling

Version 2.6
* Add -f compat flag to groupadd
* Add UTF8/local locale translation
* Make regex for allowed accounts/groups configurable
* Add optional pam_rpasswd module
* Add SLP support to rpasswd/rpasswdd
* Add pwck
* Add grpck
* Add newgrp
* Add /etc/default/passwd support

Version 2.5.1
* Install all manual pages

Version 2.5
* Adjust for libselinux v1.5
* Add pwconv/pwunconv
* Add grpunconv

Version 2.4
* Fix -G handling of usermod
* Fix LDAP support
* Add gpasswd, groupmod and usermod
* Copy always all extended attributes
* Add SELinux support for passwd, chsh and chfn
* Make sure that group, passwd and shadow are always created with
  correct security attributes
* groupadd -g handling fixed
* vipw/vigr added
* Lot of bug fixes
* Add manual pages
* Add groupadd and groupdel
* useradd and userdel can now create/delete LDAP accounts
* Make compile with OpenLDAP 2.0.x again
* Rewrite ldap functions to meet demands
* chpasswd can now change LDAP passwords
* useradd: Lot of new error checks and bug fixes
* chfn/chsh: Allow user to abort with Ctrl-C during data input
* Add test suite

Version 2.3
* Add chpasswd
* Add expiry
* Fix chage options
* rpasswd client now checks server certificate

Version 2.2
* Add "-D binddn" option, to allow modifying LDAP data with a special
  account.
* passwd and chage: Allow modifying of password expiry informations
  stored in a LDAP database.

Version 2.1
* Add --disable-ldap option to configure
* Fix typos in chage.1 manual page
* Rewrite passwd manual page and explain all options

Version 2.0
* Fix -d and -E option of chage
* Use port 774 as new default
* Make compile with glibc 2.3 and gcc 2.95/3.2/3.3

Version 1.99
* Add chage
* Allow modifying of passwd/shadow file not in /etc (only chage yet)

Version 1.98
* Add manual pages for rpasswd and rpasswdd

Version 1.97
* rpasswd now reads config file with server name and port number
* Security fix for rpasswdd: don't allow remote changing of root
  password without authentification
* Add manual page for rpasswd.conf

Version 1.96
* Make compile with gcc 2.x
* root can now change passwords for other user
* timeout if client does not answer fast enough
* Sent LANG as locale to server

Version 1.95
* Make location of ldap config files compile time configurable

Version 1.94
* Add more sanity checks to prevent file corruption

Version 1.93
* Minor bug fixes to rpasswdd

Version 1.92

src/pam_rpasswd.c
/*
 * Copyright (c) 2004, 2005, 2009, 2010 Thorsten Kukuk <kukuk@suse.de>
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, and the entire permission notice in its entirety,
 *    including the disclaimer of warranties.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. The name of the author may not be used to endorse or promote
 *    products derived from this software without specific prior
 *    written permission.
 *
 * ALTERNATIVELY, this product may be distributed under the terms of
 * the GNU Public License, in which case the provisions of the GPL are
 * required INSTEAD OF the above restrictions.  (This clause is
 * necessary due to a potential bad interaction between the GPL and
 * the restrictions contained in a BSD-style copyright.)
 *
 * THIS SOFTWARE IS PROVIDED ``AS IS'' AND ANY EXPRESS OR IMPLIED
 * WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
 * OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT,
 * INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
 * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
 * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
 * STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED
 * OF THE POSSIBILITY OF SUCH DAMAGE.
 */

#if defined(HAVE_CONFIG_H)
#include "config.h"
#endif

#include <errno.h>
#include <stdio.h>
#include <stdlib.h>
#include <stdarg.h>
#include <unistd.h>
#include <pwd.h>
#include <shadow.h>
#include <dlfcn.h>
#include <time.h>
#include <fcntl.h>
#include <syslog.h>
#include <sys/stat.h>
#include <sys/time.h>
#include <rpc/types.h>

#ifdef USE_GNUTLS
#include <gnutls/gnutls.h>
#else
#include <openssl/crypto.h>
#include <openssl/x509.h>
#include <openssl/pem.h>
#include <openssl/ssl.h>
#include <openssl/err.h>
#endif

#define PAM_SM_PASSWORD
#include <security/_pam_macros.h>
#include <security/pam_modules.h>

#include "error_codes.h"
#include "rpasswd-client.h"

/* We don't use translated texts.  */
#define _(s) s

struct options_t {
  int debug;
  int use_slp;
  int verbose;
  int quiet;
  int reqcert;
  char *config_file;
  char *host;
  char *port;
  char *descr;
};
typedef struct options_t options_t;

/* syslogging function for errors and other information */
static void
__pam_log (int err, const char *format,...)
{
  va_list args;
  char *str;

  va_start (args, format);
  if (vasprintf (&str, format, args) < 0)
    return;
  syslog (err, "pam_rpasswd: %s", str);
  va_end (args);
}

/* write message to user */
static int
__write_message (pam_handle_t *pamh, int msg_style,
                 const char *fmt,...)
{
  struct pam_message msg[1], *pmsg[1];
  struct pam_response *resp=NULL;
  struct pam_conv *conv;
  void *conv_void;
  char buffer[512];
  va_list ap;
  int retval;
  int buflen;

  va_start (ap, fmt);
  vsnprintf (buffer, sizeof (buffer), fmt, ap);
  va_end (ap);

  buflen = strlen (buffer);
  if (buffer[buflen - 1] == '\n')
    buffer[buflen - 1] = '\0';

  pmsg[0] = &msg[0];
  msg[0].msg_style = msg_style;
  msg[0].msg = buffer;

  retval = pam_get_item (pamh, PAM_CONV, (const void **) &conv_void);
  conv = (struct pam_conv *) conv_void;
  if (retval == PAM_SUCCESS)
    {
      retval = conv->conv (1, (const struct pam_message **)pmsg,
                           &resp, conv->appdata_ptr);
      if (retval != PAM_SUCCESS)
        return retval;
    }
  else
    return retval;

  msg[0].msg = NULL;
  if (resp)
    _pam_drop_reply(resp, 1);

  return retval;
}

#define HANDLE pam_handle_t *pamh,
#define IAM_PAM_MODULE
#define PRINTF __write_message
#define ERR_HANDLE pamh, PAM_ERROR_MSG
#define STD_HANDLE pamh, PAM_TEXT_INFO

#include "rpasswd-client.c"

#define CONV_ECHO_ON  1		/* types of echo state */
#define CONV_ECHO_OFF 0

static int
read_string (pam_handle_t *pamh, int echo, const char *prompt,
	     char **retstr)
{
  struct pam_message msg[1], *pmsg[1];
  struct pam_response *resp;
  struct pam_conv *conv;
  void *conv_void;
  int retval;

  /* set up conversation call */

  pmsg[0] = &msg[0];
  if (echo == CONV_ECHO_ON)
    msg[0].msg_style = PAM_PROMPT_ECHO_ON;
  else
    msg[0].msg_style = PAM_PROMPT_ECHO_OFF;
  msg[0].msg = prompt;
  resp = NULL;

  retval = pam_get_item (pamh, PAM_CONV, (const void **) &conv_void);
  conv = (struct pam_conv *) conv_void;
  if (retval == PAM_SUCCESS)
    {
      retval = conv->conv (1, (const struct pam_message **)pmsg,
                           &resp, conv->appdata_ptr);
      if (retval != PAM_SUCCESS)
        return retval;
    }
  else
    return retval;

  if (resp)
    {
      *retstr = strdup (resp->resp ? resp->resp : "");
      if (resp)
        _pam_drop_reply (resp, 1);
    }
  else
    return PAM_CONV_ERR;

  return PAM_SUCCESS;
}

#ifdef USE_GNUTLS
static int
handle_responses (pam_handle_t *pamh, gnutls_session ssl)
{
  response_header resp;
  char retval = PAM_SUCCESS;
  char *buf = NULL;

  do
    {
      int ret;

      /* header */
      if ((ret = gnutls_record_recv (ssl, &resp, sizeof (resp))) <= 0)
	{
	  if (ret == 0)
	    __write_message (pamh, PAM_ERROR_MSG,
			     _("error while reading request: %s"),
			     _("Peer has closed the TLS connection"));
	  else
	    __write_message (pamh, PAM_ERROR_MSG,
			     _("error while reading request: %s"),
			     gnutls_strerror (ret));
	  return PAM_SYSTEM_ERR;
	}

      /* first entry */
      if (resp.data_len > 0)
	{
	  buf = malloc (resp.data_len);
	  if (buf == NULL)
	    {
	      __write_message (pamh, PAM_ERROR_MSG,
			       _("error while allocating memory: %m"));
	      return PAM_SYSTEM_ERR;
	    }

	  if ((ret = gnutls_record_recv (ssl, buf, resp.data_len)) <= 0)
	    {
	      __write_message (pamh, PAM_ERROR_MSG,
			       _("error while reading request data: %s"),
			       gnutls_strerror (ret));
	      free (buf);
	      return PAM_SYSTEM_ERR;
	    }
	}

      switch (resp.type)
	{
	case TEXT_INFO:
	  if (buf)
	    __write_message (pamh, PAM_TEXT_INFO,
			     "%s\n", buf);
	  break;
	case ERROR_MSG:
	  if (buf)
	    __write_message (pamh, PAM_ERROR_MSG,
			     "%s\n", buf);
	  break;
	case PROMPT_ECHO_OFF:
	  {
	    char *string = NULL;
	    int nc = read_string (pamh, CONV_ECHO_OFF, buf, &string);
	    if (nc < 0)
	      retval = send_string (pamh, ssl, PAM_CONV_ERR, string);
	    else
	      retval = send_string (pamh, ssl, PAM_SUCCESS, string);
	  }
	  break;
	case PROMPT_ECHO_ON:
	  {
	    char *string = NULL;
	    int nc = read_string (pamh, CONV_ECHO_ON, buf, &string);
	    if (nc < 0)
	      retval = send_string (pamh, ssl, PAM_CONV_ERR, string);
	    else
	      retval = send_string (pamh, ssl, PAM_SUCCESS, string);
	  }
	  break;
	case FINISH:
	  if (buf)
	    retval = buf[0];
	  else
	    retval = PAM_SYSTEM_ERR;
	  break;
	default:
	  break;
	}

      if ((resp.type == PROMPT_ECHO_ON || resp.type == PROMPT_ECHO_OFF) &&
	  retval != 0)
	{
	  char err_buf[256];

	  PRINTF (ERR_HANDLE, _("Cannot send input back to server: %s\n"),
		  strerror_r (errno, err_buf, sizeof (err_buf)));
	  free (buf);
	  return PAM_SYSTEM_ERR;
	}
      if (buf)
	{
	  free (buf);
	  buf = NULL;
	}
    }
  while (resp.type != FINISH);

  return retval;
}
#else

static int
handle_responses (pam_handle_t *pamh, SSL *ssl)
{
  response_header resp;
  char retval = PAM_SUCCESS;
  char *buf = NULL;

  do
    {
      errno = 0;
      if (TEMP_FAILURE_RETRY (SSL_read (ssl, &resp, sizeof (resp)))
	  != sizeof (resp))
	{
	  char err_buf[256];

	  if (errno == 0)
	    __write_message (pamh, PAM_ERROR_MSG,
			     _("error while reading request: %s"),
			     _("wrong data received"));
	  else
	    __write_message (pamh, PAM_ERROR_MSG,
			     _("error while reading request: %s"),
			     strerror_r (errno, err_buf, sizeof (err_buf)));
	  return PAM_SYSTEM_ERR;
	}

      if (resp.data_len > 0)
	{
	  if (buf == NULL)
	    {
	      __write_message (pamh, PAM_ERROR_MSG,
			       _("error while allocating memory: %m"));
	      return PAM_SYSTEM_ERR;
	    }

	  if (TEMP_FAILURE_RETRY (SSL_read (ssl, buf, resp.data_len))
	      != resp.data_len)
	    {
	      __write_message (pamh, PAM_ERROR_MSG,
			       _("error while reading request data: %m"));
	      free (buf);
	      return PAM_SYSTEM_ERR;
	    }
	}

      switch (resp.type)
	{
	case TEXT_INFO:
	  if (buf)
	    __write_message (pamh, PAM_TEXT_INFO,
			     "%s\n", buf);
	  break;
	case ERROR_MSG:
	  if (buf)
	    __write_message (pamh, PAM_ERROR_MSG,
			     "%s\n", buf);
	  break;
	case PROMPT_ECHO_OFF:
	  {
	    char *string = NULL;
	    int nc = read_string (pamh, CONV_ECHO_OFF, buf, &string);
	    if (nc < 0)
	      retval = send_string (ssl, PAM_CONV_ERR, string);
	    else
	      retval = send_string (ssl, PAM_SUCCESS, string);
	  }
	  break;
	case PROMPT_ECHO_ON:
	  {
	    char *string = NULL;
	    int nc = read_string (pamh, CONV_ECHO_ON, buf, &string);
	    if (nc < 0)
	      retval = send_string (ssl, PAM_CONV_ERR, string);
	    else
	      retval = send_string (ssl, PAM_SUCCESS, string);
	  }
	  break;
	case FINISH:
	  if (buf)
	    retval = buf[0];
	  else
	    retval = PAM_SYSTEM_ERR;
	  break;
	default:
	  break;
	}

      if ((resp.type == PROMPT_ECHO_ON || resp.type == PROMPT_ECHO_OFF) &&
	  retval != 0)
	{
	  PRINTF (ERR_HANDLE, _("Cannot send input back to server: %m\n"));
	  free (buf);
	  return PAM_SYSTEM_ERR;
	}

      if (buf)
	{
	  free (buf);
	  buf = NULL;
	}
    }
  while (resp.type != FINISH);

  return retval;
}
#endif

static void
parse_option (const char *argv, options_t *options)
{
  if (argv == NULL || argv[0] == '\0')
    return;

  if (strcasecmp (argv, "debug") == 0)
    options->debug = 1;
  else if (strcasecmp (argv, "use_slp=1") == 0)
    options->use_slp = 1;
  else if (strcasecmp (argv, "use_slp=0") == 0)
    options->use_slp = 0;
  else if (strcasecmp (argv, "verbose") == 0)
    options->verbose = 1;
  else if (strcasecmp (argv, "quiet") == 0)
    options->quiet = 1;
  else if (strncasecmp (argv, "reqcert=", 8) == 0)
    options->reqcert = parse_reqcert (&argv[8]);
  else if (strncasecmp (argv, "config_file=", 12) == 0)
    options->config_file = strdup (&argv[12]);
  else if (strncasecmp (argv, "host=", 5) == 0)
    options->host = strdup (&argv[5]);
  else if (strncasecmp (argv, "port=", 5) == 0)
    options->port = strdup (&argv[5]);
  else
    __pam_log (LOG_ERR, "Unknown option: `%s'", argv);
}


static options_t *
get_options (int argc, const char **argv)
{
  options_t *options = calloc (1, sizeof (options_t));

  if (options == NULL)
    return NULL;

  options->reqcert = 3;
  options->config_file = _PATH_RPASSWDCONF;

  /* Parse parameters for module */
  for ( ; argc-- > 0; argv++)
    parse_option (*argv, options);

  if (options->quiet)
    options->verbose = 0;

  return options;
}

PAM_EXTERN int
pam_sm_chauthtok (pam_handle_t *pamh, int flags,
		  int argc, const char **argv)
{
  options_t *options;
  void *user_void;
  char *user;
  int retval;
  int sock;
#ifdef USE_GNUTLS
  gnutls_session session;
  gnutls_certificate_credentials xcred;
#else
  SSL_CTX *ctx;
  SSL *ssl;
#endif
#ifdef USE_SLP
  int used_slp = 0;
#endif

  options = get_options (argc, argv);
  if (options == NULL)
    {
      __pam_log (LOG_ERR, "cannot get options");
      return PAM_BUF_ERR;
    }

  if (options->debug)
    __pam_log (LOG_DEBUG, "pam_sm_chauthtok() called");


  retval = pam_get_item (pamh, PAM_USER, (const void **) &user_void);
  if (retval != PAM_SUCCESS)
    {
      free (options);
      return retval;
    }
  user = (char *) user_void;

  if (user == NULL || strlen (user) == 0)
    {
      if (options->debug)
	__pam_log (LOG_DEBUG, "user (%s) unknown", user ? user : "NULL");
      /* The app is supposed to get us the username! */
      free (options);
      return PAM_USER_UNKNOWN;
    }

  if (flags & PAM_PRELIM_CHECK)
    {
      free (options);
      return PAM_SUCCESS;
    }

  if (options->host)
    {
      if (options->port)
        load_config (options->config_file, NULL, NULL, &options->reqcert);
      else
        load_config (options->config_file, NULL, &options->port,
		     &options->reqcert);
    }
  else if (options->port)
    load_config (options->config_file, &options->host, NULL,
		 &options->reqcert);
  else
    load_config (options->config_file, &options->host,
		 &options->port, &options->reqcert);


#ifdef USE_SLP
  if (options->host == NULL && options->port == NULL &&
      options->use_slp == 1)
    {
      query_slp (pamh, &options->host, &options->port, &options->descr);
      used_slp = 1;
    }
#endif

  if (options->host == NULL)
    {
      PRINTF (ERR_HANDLE, "No server specified\n");
      free (options);
      return PAM_SYSTEM_ERR;
    }

  if (options->port == NULL)
    options->port = "rpasswd";

#ifdef USE_SLP
  if (used_slp)
    {
      if (options->port != NULL && strcmp (options->port, "rpasswd") != 0)
	{
	  if (options->descr)
	    PRINTF (STD_HANDLE, _("SLP: Found Server on %s, port %s (%s)\n"),
		    options->host, options->port, options->descr);
	  else
	    PRINTF (STD_HANDLE, _("SLP: Found Server on %s, port %s\n"),
		    options->host, options->port);
	}
      else
	{
	  if (options->descr)
	    PRINTF (STD_HANDLE, _("SLP: Found Server on %s (%s)\n"),
		    options->host, options->descr);
	  else
	    PRINTF (STD_HANDLE, _("SLP: Found Server on %s\n"),
		    options->host);
	}
    }
#endif

  /* Now we have all the initial information we need from the app to
     set things up (we assume that getting the username succeeded...) */
  sock = connect_to_server (pamh, options->host, options->port,
			    PF_UNSPEC, options->quiet);
  if (sock < 0)
    {
      free (options);
      return PAM_TRY_AGAIN;
    }

  /* Do SSL */
#ifdef USE_GNUTLS
  gnutls_global_init ();
  /* Initialize TLS session. */
  gnutls_init (&session, GNUTLS_CLIENT);
  retval = start_ssl (pamh, sock, options->reqcert, options->verbose,
		      session, &xcred);
#else
  retval = start_ssl (pamh, sock, options->reqcert, options->verbose,
		      &ctx, &ssl);
#endif

  if (retval != 0)
    {
      free (options);
      return PAM_SYSTEM_ERR;
    }

#ifdef USE_GNUTLS
  if ((retval = start_request (pamh, session, user, 0)) == 0)
    retval = handle_responses (pamh, session);
  else
    retval = PAM_SYSTEM_ERR;
#else
  if ((retval = start_request (ssl, user, 0)) == 0)
    retval = handle_responses (pamh, ssl);
  else
    retval = PAM_SYSTEM_ERR;
#endif

#ifdef USE_GNUTLS
  gnutls_bye (session, GNUTLS_SHUT_RDWR);
  close (sock);
  gnutls_deinit (session);
  gnutls_certificate_free_credentials (xcred);
  gnutls_global_deinit ();
#else
  close (sock);
  SSL_free (ssl);
  SSL_CTX_free (ctx);
#endif

  free (options);

  return retval;
}
